### PR TITLE
Added g2p (grapheme to phoneme) problem

### PR DESCRIPTION
**Pull request by nshmyrev**
G2P (grapheme to phoneme) is a problem to convert the written word to phonemes, a very popular task in speech recognition. This initial PR adds a problem description based on CMUDict dataset.

### CODE REVIEW

**Comment by nshmyrev** [comment]
I signed it!

File: tensor2tensor/data_generators/g2p.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    vocab_path: path where to save the vocabulary.
+  """
+  vocab = set()
+  with tf.gfile.GFile(filename, "r") as f:
+    for line in f:
+        items = line.split()
+        vocab.update(items[0])
+        vocab.update(items[1:])
+  with open(vocab_path, "w") as f:
+    f.write("<pad>\n")
```
Comment (rsepassi) [inline tensor2tensor/data_generators/g2p.py]: 2 space indents

Comment (rsepassi) [inline tensor2tensor/data_generators/g2p.py]: use the constants defined in `text_encoder`

File: tensor2tensor/data_generators/g2p.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    f.write("<EOS>\n")
+    f.write("\n".join(sorted(vocab)))
```
Comment (rsepassi) [inline tensor2tensor/data_generators/g2p.py]: use constant

File: tensor2tensor/data_generators/g2p.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+from tensor2tensor.utils import registry
+
+import tensorflow as tf
+
+EOS = text_encoder.EOS
+EOS_ID = text_encoder.EOS_ID
+CMUDICT_URL = "https://sourceforge.net/projects/cmusphinx/files/G2P%20Models/phonetisaurus-cmudict-split.tar.gz/download"
+
+
+def _build_vocab(filename, vocab_path):
```
Comment (rsepassi) [inline tensor2tensor/data_generators/g2p.py]: TokenTextEncoder can be initialized with a `vocab_list` and has a `store_to_file` method. Could you use those instead and verify that the first 2 entries in the saved file are PAD and EOS?

File: tensor2tensor/data_generators/g2p.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+  def has_inputs(self):
+    return True
+
+  @property
+  def target_space_id(self):
+    return problem.SpaceID.EN_TOK
+
+  @property
+  def input_space_id(self):
+    return problem.SpaceID.EN_TOK
```
Comment (rsepassi) [inline tensor2tensor/data_generators/g2p.py]: are the inputs and outputs actually english tokens? if so, these are fine. if not, see if there's a space id that describes the data, or use GENERIC.

File: tensor2tensor/data_generators/g2p.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+  @property
+  def num_shards(self):
+    return 1
```
Comment (rsepassi) [inline tensor2tensor/data_generators/g2p.py]: how big is the dataset?

File: tensor2tensor/data_generators/g2p.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  def vocab_name(self):
+    return "vocab.g2p"
+
+  @property
+  def use_subword_tokenizer(self):
+    return False
+
+  @property
+  def targeted_vocab_size(self):
+    return 10000
```
Comment (rsepassi) [inline tensor2tensor/data_generators/g2p.py]: `_build_vocab` doesn't seem to enforce a size at all, so this seems misleading. if the vocab is supposed to fully span the dataset and there would never be any real reason to change the vocab size, then you can override `vocab_file` directly and have it just be `"vocab.g2p"`

File: tensor2tensor/data_generators/g2p.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+def _get_token_encoder(vocab_dir, vocab_name, filename):
+  """Reads from file and returns a `TokenTextEncoder` for the vocabulary."""
+  vocab_path = os.path.join(vocab_dir, vocab_name)
+  if not tf.gfile.Exists(vocab_path):
+    _build_vocab(filename, vocab_path)
+  return text_encoder.TokenTextEncoder(vocab_path)
+
+@registry.register_problem
+class G2p(problem.Text2TextProblem):
```
Comment (rsepassi) [inline tensor2tensor/data_generators/g2p.py]: this name seems a bit too short for comfort. Grapheme2Phoneme, or GraphemeToPhoneme

File: tensor2tensor/data_generators/g2p.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    filename = os.path.basename(CMUDICT_URL)
+    compressed_filepath = generator_utils.maybe_download(
+        tmp_dir, filename, CMUDICT_URL)
+
+    with tarfile.open(compressed_filepath, "r:gz") as tgz:
+      files = [m.name for m in tgz.getmembers()]
+      tgz.extractall(tmp_dir)
+
+    train_file, valid_file = None, None
+    for filename in files:
```
Comment (rsepassi) [inline tensor2tensor/data_generators/g2p.py]: could you assert that `len(files) == 2`

**Comment by nshmyrev** [comment]
Sorry, holiday time. I'll try to update the pull request soon.

**Comment by lukaszkaiser** [comment]
@nshmyrev : are you back? We'd love to merge it, but some changes are needed on your side. Let us know!

**Comment by lukaszkaiser** [comment]
Closing for now due to lack of activity.

**Comment by loretoparisi** [comment]
@lukaszkaiser @nshmyrev so this has not been merged at the end?

**Comment by nshmyrev** [comment]
I still hope to find the time to work on this, maybe this month.

**Pull request closed without merging**